doc/api_ref/pubkey.rst

```diff
@@ -1089,34 +1279,34 @@ produce an output of the desired length.
       Set up to perform key derivation using the given private key and specified KDF.
 
   .. cpp:function:: SymmetricKey derive_key(size_t key_len, \
-                    const uint8_t in[], \
-                    size_t in_len, \
-                    const uint8_t params[], \
-                    size_t params_len) const
+                    const uint8_t peer_key[], \
+                    size_t peer_key_len, \
+                    const uint8_t salt[], \
+                    size_t salt_len) const
 
   .. cpp:function:: SymmetricKey derive_key(size_t key_len, \
-                    std::span<const uint8_t> in, \
-                    const uint8_t params[], size_t params_len) const
+                    std::span<const uint8_t> peer_key, \
+                    const uint8_t salt[], size_t salt_len) const
 
   .. cpp:function:: SymmetricKey derive_key(size_t key_len, \
-                    const uint8_t in[], size_t in_len, \
-                    const std::string& params = "") const
+                    const uint8_t peer_key[], size_t peer_key_len_len, \
+                    const std::string& salt = "") const
 
   .. cpp:function:: SymmetricKey derive_key(size_t key_len, \
-                    const std::span<const uint8_t> in, \
-                    const std::string& params = "") const
+                    std::span<const uint8_t> peer_key, \
+                    const std::string& salt = "") const
 
-     Return a shared key. The *params* will be hashed along with the shared secret by the
+     Return a shared key. The *salt* will be hashed along with the shared secret by the
      KDF; this can be useful to bind the shared secret to a specific usage.
```

Comment:
... if no KDF was defined, the salt and the key length are typically ignored. (Or a formulation along those lines.)

I actually find this to be an API pitfall, because users might have designed their protocol around the idea of injecting that context while it fact it is just silently ignored. Here, a to-be-built API along the lines of `PK_Options` (#4318) might come in handy. It would make it easy to throw an exception if `.with_salt()` (or `.with_output_length()`) was given but no KDF actually ended up consuming those values in `derive_key()`.

Reply:
Actually we already do reject a salt if a KDF is not being used but I agree with your point in the broad sense.

Reply:
Updated to describe how keylen param is ignored without a kdf